pkg/controller/certificates/rootcacertpublisher/publisher_test.go

```diff
@@ -76,134 +75,102 @@ func TestConfigMapCreation(t *testing.T) {
-		if err != nil {
-			t.Fatalf("error creating ServiceAccounts controller: %v", err)
-		}
-		controller.cmListerSynced = alwaysReady
-		controller.nsListerSynced = alwaysReady
-
-		cmStore := cmInformer.Informer().GetStore()
-		nsStore := nsInformer.Informer().GetStore()
-
-		syncCalls := make(chan struct{})
-		controller.syncHandler = func(key string) error {
-			err := controller.syncNamespace(key)
+		t.Run(k, func(t *testing.T) {
+			client := fake.NewSimpleClientset(caConfigMap, existNS)
+			informers := informers.NewSharedInformerFactory(fake.NewSimpleClientset(), controller.NoResyncPeriodFunc())
+			cmInformer := informers.Core().V1().ConfigMaps()
+			controller, err := NewPublisher(cmInformer, client, fakeRootCA)
 			if err != nil {
-				t.Logf("%s: %v", k, err)
+				t.Fatalf("error creating ServiceAccounts controller: %v", err)
+			}
+
+			cmStore := cmInformer.Informer().GetStore()
+
+			controller.syncHandler = controller.syncNamespace
+
+			for _, s := range tc.ExistingConfigMaps {
+				cmStore.Add(s)
+			}
+
+			if tc.AddedNamespace != nil {
+				controller.namespaceAdded(tc.AddedNamespace)
+			}
+			if tc.UpdatedNamespace != nil {
+				controller.namespaceUpdated(nil, tc.UpdatedNamespace)
 			}
-			syncCalls <- struct{}{}
-			return err
-		}
-		stopCh := make(chan struct{})
-		defer close(stopCh)
-		go controller.Run(1, stopCh)
-
-		if tc.ExistingNamespace != nil {
-			nsStore.Add(tc.ExistingNamespace)
-		}
-		for _, s := range tc.ExistingConfigMaps {
-			cmStore.Add(s)
-		}
-
-		if tc.AddedNamespace != nil {
-			nsStore.Add(tc.AddedNamespace)
-			controller.namespaceAdded(tc.AddedNamespace)
-		}
-		if tc.UpdatedNamespace != nil {
-			controller.namespaceUpdated(tc.ExistingNamespace, tc.UpdatedNamespace)
-		}
-
-		if tc.DeletedConfigMap != nil {
-			cmStore.Delete(tc.DeletedConfigMap)
-			controller.configMapDeleted(tc.DeletedConfigMap)
-		}
-
-		if tc.UpdatedConfigMap != nil {
-			old := tc.UpdatedConfigMap[0]
-			new := tc.UpdatedConfigMap[1]
-			controller.configMapUpdated(old, new)
-		}
-
-		// wait to be called
-		select {
-		case <-syncCalls:
-		case <-time.After(5 * time.Second):
-		}
-
-		actions := client.Actions()
-		if len(tc.ExpectActions) != len(actions) {
-			t.Errorf("%s: Expected to create configmap %#v. Actual actions were: %#v", k, tc.ExpectActions, actions)
-			continue
-		}
-		for i, expectAction := range tc.ExpectActions {
-			action := actions[i]
-			if !action.Matches(expectAction.verb, "configmaps") {
-				t.Errorf("%s: Unexpected action %s", k, action)
-				break
+
+			if tc.DeletedConfigMap != nil {
+				cmStore.Delete(tc.DeletedConfigMap)
+				controller.configMapDeleted(tc.DeletedConfigMap)
+			}
+
+			if tc.UpdatedConfigMap != nil {
+				cmStore.Add(tc.UpdatedConfigMap)
+				controller.configMapUpdated(nil, tc.UpdatedConfigMap)
+			}
+
+			for controller.queue.Len() != 0 {
+				controller.processNextWorkItem()
+			}
+
+			actions := client.Actions()
+			if len(tc.ExpectActions) != len(actions) {
```

Comment:
Do we already use https://godoc.org/github.com/google/go-cmp/cmp in k8s? It would be a bit nicer than manually comparing slices